network/p2p/dns/resolver.go

```diff
@@ -158,19 +221,12 @@ func (r *Resolver) lookupTXT(ctx context.Context, txt string) ([]string, error)
 		return r.lookupResolverForTXTAddr(ctx, txt)
 	}
 
-	if !fresh && r.shouldResolveTXT(txt) {
-		r.unit.Launch(func() {
-			defer r.doneResolvingTXT(txt)
-			_, err := r.lookupResolverForTXTAddr(ctx, txt)
-			if err != nil {
-				// invalidates cached entry when hits error on resolving.
-				invalidated := r.c.invalidateTXTCacheEntry(txt)
-				if invalidated {
-					r.collector.OnDNSCacheInvalidated()
-				}
-			}
-		})
-
+	if !fresh && r.shouldResolveTXT(txt) && !util.CheckClosed(r.cm.ShutdownSignal()) {
+		select {
+		case r.txtRequests <- &lookupTXTRequest{txt}:
+		default:
+			r.logger.Warn().Str("txt", txt).Msg("TXT lookup request queue is full, dropping request")
+		}
```

Comment:
🚩  This part is needed to get refactored to make sure updating an expired DNS request is not consigned to starvation, i.e., Each time the DNS cache is queried for a domain that is expired, the DNS cache is out of available workers, and hence keeps only returning that expired entry. This can simply happen if the DNS query rate outpaces the number of workers' resolution rate. The previous approach was taking care of this edge case. You may revert to the previous version (i.e., `engine.Unit`-based one) if resolving this edge case is not trivial.